source-mongodb/pull.go

```diff
@@ -132,141 +109,213 @@ func (d *driver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) err
 		}
-		var startAt = &primitive.Timestamp{T: uint32(state.BackfillStartedAt.Unix()) - 1, I: 0}
+		var startAt = &primitive.Timestamp{T: uint32(startAtWithSafety.Unix()) - 1, I: 0}
 		opts = opts.SetStartAtOperationTime(startAt)
 	}
 
-	cursor, err := collection.Watch(ctx, mongo.Pipeline{eventFilter}, opts)
+	// First attempt to do an instance-wide change stream, this is possible if we
+	// have readAnyDatabase access on the instance. If this fails for permission
+	// issues, we then attempt to do a database-level change stream. In the case
+	// of the database-level change stream we expect only a single database to be
+	// provided and do not support multiple databases in this mode.
+	cursor, err := client.Watch(ctx, mongo.Pipeline{eventFilter}, opts)
 	if err != nil {
 		// This error means events from the starting point are no longer available,
 		// which means we have hit a gap in between last run of the connector and
-		// this one. We re-set the checkpoint and error out so the next run of the
-		// connector will run a backfill.
-		//
-		// There is potential for optimisation here by only resetting the checkpoint
-		// of the collection which can't resume instead of resetting the whole
-		// connector's checkpoint and backfilling everything
+		// this one.
 		if e, ok := err.(mongo.ServerError); ok {
 			if e.HasErrorMessage(resumePointGoneErrorMessage) || e.HasErrorMessage(resumeTokenNotFoundErrorMessage) {
-				return fmt.Errorf("change stream on collection %s cannot resume capture, this is usually due to a small oplog storage available. Please resize your oplog to be able to safely capture data from your database: https://go.estuary.dev/NurkrE. After resizing your uplog, you can remove the binding for this collection add it back to trigger a backfill: %w", res.Collection, e)
+				cp, err := json.Marshal(captureState{})
+				if err != nil {
+					return fmt.Errorf("marshalling checkpoint for removed bindings: %w", err)
+				} else if err = c.Output.Checkpoint(cp, false); err != nil {
+					return fmt.Errorf("updating checkpoint for removed bindings: %w", err)
+				}
+				return fmt.Errorf("change stream cannot resume capture, this is usually due to insufficient oplog storage. Please consider resizing your oplog to be able to safely capture data from your database: https://go.estuary.dev/NurkrE. The connector will now attempt to backfill all bindings again: %w", e)
+			} else if e.HasErrorMessage(unauthorizedMessage) {
+				log.WithField("err", e).Warn("could not start instance-level change stream, trying database-level change stream.")
+				// Check that there is only one database, if so, we can start a change
+				// stream on this database. If there is more than one database
+				// configured, we error out as we do not support multiple change streams
+				// on different databases. An instance-level change stream must be used
+				// for that scenario
+				var dbName = resources[0].Database
+				for _, res := range resources[1:] {
```

Comment:
Does this work if there is only a single resource?

Reply:
Yes it works: https://go.dev/play/p/fsPSvjMP8YT